Supported dynamic secret on initialization

mangchiandjjoe opened the pull request:
*title*: *one line description*

>Title can be a one or two words to describe the subsystem or the aspect
 this PR applies to. Title and description must be lower case. For example:
* ci: update build image to 44d539cb
* docs: fix indent, buffer: add copyOut() method
* router:add x-envoy-overloaded header
* tls: add support for specifying TLS session ticket keys

*Description*:
>What does this PR do? What was the behavior before the PR?
What is the behavior with the PR? If fixing a bug, please describe what
the original issue is and how the change resolves it. How does this
feature get enabled? By default, config change, etc...

*Risk Level*: Low | Medium | High
>Low: Small bug fix or small optional feature.

>Medium: New features that are not enabled(for example: new filter). Small-medium
features added to existing components(for example: modification to an existing
filter).

>High: Complicated changes such as flow control, rewrites of critical
components, etc.

>Note: The above is only a rough guide for choosing a level,
please ask if you have any concerns about the risk of the PR.

*Testing*:
>Explanation of what testing was done, for example: unit test,
integration, manual testing, etc.

>Note: It isn’t expected to do all
forms of testing, please use your best judgement or ask for guidance
if you are unsure. A good rule of thumb is the riskier the change, the
more comprehensive the testing should be.

*Docs Changes*:
>Link to [Data Plane PR](https://github.com/envoyproxy/data-plane-api/pulls)]
if your PR involves documentation changes. Please write in N/A if there were no
documentation changes.

*Release Notes*:
>If this change is user impacting you **must** add a release note to
[RAW_RELEASE_NOTES.md](RAW_RELEASE_NOTES.md). Thank you! Please write in N/A if
there are no release notes.

[Optional Fixes #Issue]

[Optional *API Changes*:]
>Link to [Data Plane PR](https://github.com/envoyproxy/data-plane-api/pulls)]

[Optional *Deprecated*:]
>Description of what is deprecated.

mattklein123 committed:
docs: hide unimplemented fields (#3448)

Signed-off-by: Matt Klein <mklein@lyft.com>

mattklein123 committed:
mongo: add adding missing BSON decoding error checking (#3444)

Signed-off-by: Matt Klein <mklein@lyft.com>

mattklein123 committed:
docs: add missing deprecation docs (#3449)

Signed-off-by: Matt Klein <mklein@lyft.com>

alyssawilk committed:
Apply whitespace checks to .proto files (#3450)

Description: makes sure proto files don't have overenthusiastic spaces

Risk Level: n/a
Testing: manually added whitespace.
Docs Changes: n/a
Release Notes: n/a
Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

jmarantz committed:
util: move the LockGuards out of the pure thread interface, and into their own library. (#3446)

Signed-off-by: Joshua Marantz <jmarantz@google.com>

Commit:
Remove the ioctl(FIONREAD) call before socket reads (#3396)

* Remove the ioctl(FIONREAD) call before socket reads

Signed-off-by: Brian Pane <bpane@pinterest.com>

mattklein123 committed:
ci: more IPv6 debugging (#3451)

Signed-off-by: Matt Klein <mklein@lyft.com>

Commit:
Supported dynamic secret on initialization

snowp committed:
server: pause rds when initializing static listeners (#3413)

In order to prevent RDS requests from being fired up before we know all
the route configurations we care about, pause RDS while we initialize
all the listeners from static config. This matches what we do during LDS
update.

We pause the RDS updates over ADS immediately, then once we know that all
the init callbacks have been completed we resume RDS updates. This was
necessary because the RDS subscriptions happen in the init callbacks.

Risk Level:
Low: Small bug fix

Testing:
Manually verified that it fixes the repro case from #3399

Docs Changes:
N/A

Release Notes:
N/A

Fixes #3399

Signed-off-by: Snow Pettersen <snowp@squareup.com>

Commit:
Listener manager and cluster manager call refresh transport_socket_factory

Commit:
Changed interfaces to const

mattklein123 committed:
governance: add details on extension policy (#3460)

Fixes https://github.com/envoyproxy/envoy/issues/2882

Signed-off-by: Matt Klein <mklein@lyft.com>

akonradi committed:
router: make IP hash policy ignore downstream port (#3457)

Downstream clients initiate connections using ephemeral ports, which are
included in the Address::Instance::asString implementation. Including
the ports breaks client IP hashing for repeated short-lived connections,
though long-lived HTTP2 connections are unaffected. This patch changes
the IP hash policy to ignore ports and hash only on the IP address.

Signed-off-by: Alex Konradi <akonradi@google.com>

jmarantz committed:
util: Add Thread::CondVar with thread annotations, and use it in source/... (#3453)

Signed-off-by: Joshua Marantz <jmarantz@google.com>

mangchiandjjoe committed:
Merge branch 'sds_dynamic_secret_merge' into sds_dynamic_secret

lizan committed:
docs: transocder route config (#3461)

Signed-off-by: Lizan Zhou <zlizan@google.com>

Commit:
Fixed compilation error

PiotrSikora committed:
listener: rename "sni_domains" to "server_names". (#3454)

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

objectiser committed:
tracing: Support 128bit trace id with zipkin tracer (#3386)

Signed-off-by: Gary Brown <gary@brownuk.com>

PiotrSikora committed:
tls: add support for multiple verify_certificate_hash values. (#3437)

While there, add support for the standard hex-encoded SHA-256 hashes without colon delimiters.

Risk Level: Low
Testing: Unit tests added.
Docs Changes: Added
Release Notes: Added

Fixes #3418, #3419.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

akonradi committed:
routing: allow generating session cookies (#3462)

This enables configuring Envoy to generate cookies that expire at the
end of a session instead of requiring them to have an explicit max-age.

Risk Level: Low

Testing: added unit tests and an integration test

Docs Changes: documented new behavior in API and release docs

Release Notes: router: allow cookie routing to generate session cookies.

Signed-off-by: Alex Konradi <akonradi@google.com>

PiotrSikora committed:
listener: perform server name match against all wildcard domains. (#3467)

*Risk Level*: Low
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

Fixes #3363.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

ramaraochavali committed:
stats:handle main thread in merge (#3445)

Signed-off-by: Rama <rama.rao@salesforce.com>

mattklein123 committed:
format: run clang-format on proto files (#3473)

Signed-off-by: Matt Klein <mklein@lyft.com>

PiotrSikora committed:
common: don't throw from Hex::decode(). (#3468)

This matches behavior of Base64::decode() and allows for a more complex
result validation to be performed by the caller.

*Risk Level*: Low
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

Fixes #3442.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

akonradi committed:
cleanup: code cleanup, no functionality changes (#3474)

Signed-off-by: Alex Konradi <akonradi@google.com>

mattklein123 committed:
Delete sync.sh (#3476)

Signed-off-by: Matt Klein <mklein@lyft.com>

rodaine committed:
authz: RBAC filter config PBs + flexibility changes (#3477)

Signed-off-by: Chris Roche <croche@lyft.com>

cpakulski committed:
config_verify: Fixed crash when config processing fails. (#3441)

Signed-off-by: Christoph Pakulski <paker8848@gmail.com>

mangchiandjjoe committed:
All required features for SDS integration is working

PiotrSikora committed:
build: sort load() arguments. (#3483)

*Risk Level*: None
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

Found with buildifier.

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

jmarantz committed:
util: replace all std::mutex and std::condition_variable with wrapped absl equivalents. (#3464)

Signed-off-by: Joshua Marantz <jmarantz@google.com>

moderation committed:
Upgrade to gperftools 2.7, spdlog 0.17.0, grpc 1.12.0, opentracing-cpp 1.4.2, jinja 2.10, rules_go 0.11.1. (#3488)

Signed-off-by: Michael Payne <michael@sooper.org>

jmarantz committed:
stats: Fix 32-bit limit for byte offsets in stats set, and guard against very high max-stats that will cause shared memory to explode. (#3480)

Signed-off-by: Joshua Marantz <jmarantz@google.com>

gumpt committed:
router: add support for inverse header matching (#3443)

Signed-off-by: Matthew Gumport <me@gum.pt>

moderation committed:
Upgrade API dependencies googleapis, gogoproto, prometheus, opencensus (#3489)

Signed-off-by: Michael Payne <michael@sooper.org>

PiotrSikora committed:
secret: add CertificateValidationContext as a possible Secret. (#3482)

No functional changes, only API update.

*Risk Level*: Low
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

cpakulski committed:
config_validate: Fix for crash when running in config validation mode with DNS specific config (#3486)

Signed-off-by: Christoph Pakulski <paker8848@gmail.com>

PiotrSikora committed:
listener: fix automatic injection of TLS Inspector. (#3490)

Previously, automatic injection was based only on the requirements
of the last filter chain and not all of them.

*Risk Level*: Low
*Testing*: bazel test //test/... (extended tests fail without the fix)
*Docs Changes*: n/a
*Release Notes*: n/a

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

dio committed:
Upgrade fmt to 4.1.0 (#3491)

This is a follow up of: #3488.

As mentioned in https://github.com/gabime/spdlog/releases/tag/v0.17.0,
spdlog v0.17.0 bumped the bundled fmt version to 4.1.0. I think Envoy
should upgrade to that version too.

Risk Level: Low

Testing: bazel test //test/...

Docs Changes: none required.

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

rodaine committed:
authz: RBAC HTTP filter + utilities (#3455)

Signed-off-by: Chris Roche <croche@lyft.com>

dnoe committed:
RBAC filter test: Check return value. (#3497)

Protobuf::util::JsonStringToMessage has a return value that shouldn't be
ignored. ASSERT_TRUE on it being OK in the RBAC filter integration test.

Testing: Ran existing tests.

Risk Level: low

Signed-off-by: Dan Noé <dpn@google.com>

jmarantz committed:
util: Adds some basic unit tests for check_format.py. (#3495)

Description:

Adds a bit of unit-testing to the format fixer/checker script, which has grown complex. These unit tests are by no means comprehensive but I thought I'd get feedback on the pattern before going much further. I actually don't have a huge amount of experience writing gunit tests in python, and it looks like there's currently no gunit tests in Python in the Envoy codebase. But I'm willing to switch this to gunit if we think that's useful.

The underlying python test program MUST be run under docker, so it's got an awkward name and comes with a bash script to run it under docker directly.

IMO check_format.py has grown organically and needs to be refactored. I have such a refactor pending, but it will be a lot easier to get confidence in the refactor when there are tests.

Once the script is tested and refactored, it will be trivial to add detection of std::mutex and std::condition_variable, which is the ultimate goal.

Risk Level: Low

Testing: tools/check_format.py -- to make sure this doesn't introduce format errors itself.
Manually ran test/check_format_test.sh.

Docs Changes: N/A
Release Notes: N/A

Signed-off-by: Joshua Marantz <jmarantz@google.com>

PiotrSikora committed:
tls: update BoringSSL to 2a52ce79 (3396). (#3507)

*Risk Level*: Low
*Testing*: bazel test //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

mattklein123 committed:
access log service: cleanup and docs (#3494)

Fixes https://github.com/envoyproxy/envoy/issues/743

This is a general cleanup of all of the access logging documentation.
I have reorganized a bunch of things and hidden the various gRPC logging
fields that are not implemented yet.

I've also moved the existing tap protos into a new "output" directory. This
is the best name I could come up for cleanly separating output data that might
be stored outside of any service or configuration.

Signed-off-by: Matt Klein <mklein@lyft.com>

snowp committed:
api: add deprecation warning to comment (#3511)

The proto field is marked as deprecated without any explanation, so this
adds a reference to the other field which should be used instead.

Signed-off-by: Snow Pettersen <snowp@squareup.com>

jmarantz committed:
stats: bump hot restart version given the data structure change. (#3506)

Signed-off-by: Joshua Marantz <jmarantz@google.com>

wozz committed:
grpc auth: add file based grpc credentials extension (#3438)

Adds a file based grpc credentials extension. See issue #3392 for more details.

Risk Level: Low: extension for grpc credentials loaded by explicit configuration options

Testing: tests included in PR

Docs Changes: Inline docs via comments and proto docs

Release Notes: N/A

Fixes #3392

Signed-off-by: Michael Wozniak <wozz@koh.ms>

lizan committed:
deps: update grpc_httpjson_transcoding (#3510)

Risk Level: Low
Testing: ran existing tests / CI
Docs Changes: N/A
Release Notes: N/A

Signed-off-by: Lizan Zhou <zlizan@google.com>

ambuc committed:
Mark FactoryContext localInfo method as const (#3514)

Signed-off-by: James Buckland <jbuckland@google.com>

mbanikazemi committed:
Adds a Lua filter example (#3513)

Adds a complete example with an echo server where a lua filter is used
to add headers to request and respons messages.

Fixes: #2152

Co-authored-by: Alex Buchanan <buchanae@gmail.com>
Co-authored-by: Robert Carosi <robert@carosi.nl>
Signed-off-by: Mohammad Banikazemi <MBanikazemi@gmail.com>

htuch committed:
hcm/router: configuration options for HTTP connection manager and router. (#3499)

HCM and router changes to support use of Envoy in scenarios where we don't want Envoy to be generating additional headers or manipulating XFF. This also introduces Via support.

Fixes #1030.

Risk Level: Low (opt in)
Testing: Unit and integration tests added.

Signed-off-by: Harvey Tuch <htuch@google.com>

lizan committed:
fix proto format (#3517)

Signed-off-by: Lizan Zhou <zlizan@google.com>

Commit:
grpc decoder: expose whether it has buffered partial data (#3524)

Signed-off-by: jiangtaoli2016 <jiangtao@google.com>

ggreenway committed:
Fix build when ENVOY_GOOGLE_GRPC is not defined. (#3534)

Signed-off-by: Greg Greenway <ggreenway@apple.com>

alyssawilk committed:
Adding Lizan as an Envoy maintainer (#3531)

Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

Commit:
gzip filter: add runtime gateway (#3527)

Signed-off-by: Lita Cho <lcho@lyft.com>

Commit:
Pull in an http_parser performance fix (#3505)

Signed-off-by: Brian Pane <bpane@pinterest.com>

mattklein123 committed:
docs: fix x-envoy-original-path docs (#3535)

Signed-off-by: Matt Klein <mklein@lyft.com>

rshriram committed:
Remove outdated doc about using deprecated-v1 (#3543)

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

zuercher committed:
mongo_proxy: fix peak/peek typo (#3541)

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

mangchiandjjoe committed:
Removed EnvoyResourceDependencyException

mangchiandjjoe committed:
Removed EnvoyResourceDependencyException

mangchiandjjoe committed:
Handle nullptr on newConnection object

alyssawilk committed:
http: making routing cookies HttpOnly, allowing configured paths (#3533)

Risk Level: Low
Testing: new unit tests
Docs Changes: config documented inline.
Release Notes: added in #3462
Fixed #3532

Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

mangchiandjjoe committed:
Cleanup deprecated codes

mangchiandjjoe committed:
Removed unecessar changes

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

mangchiandjjoe committed:
Removed unecessary changes

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

ikonst committed:
router: gRPC retry on 'unavailable' status (#3542)

Adding gRPC unavailable (14) status code to the retriable gRPC errors, e.g.
```
x-envoy-retry-grpc-on: unavailable
```

Rationale - from the Go gRPC implementation:
> Unavailable indicates the service is currently unavailable. This is a most likely a transient condition and may be corrected by retrying with a backoff.

Signed-off-by: Ilya Konstantinov <ilya.konstantinov@gmail.com>

PiotrSikora committed:
tls: add support for verify_certificate_spki. (#3475)

Signed-off-by: Piotr Sikora <piotrsikora@google.com>

mangchiandjjoe committed:
Changed the data structure of the secret update callback

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

mangchiandjjoe committed:
Fixed code format

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

lizan commented on include/envoy/network/transport_socket.h:
```diff
@@ -144,7 +145,7 @@ typedef std::unique_ptr<TransportSocket> TransportSocketPtr;
  * A factory for creating transport socket. It will be associated to filter chains and clusters.
  */
-class TransportSocketFactory {
+class TransportSocketFactory : public Secret::SecretCallbacks {
```
revert this file, not all TransportSocketFacotry need this, it can be done in impl side.

mangchiandjjoe replied:
Reverted

lizan commented on include/envoy/secret/secret_manager.h:
```diff
@@ -21,17 +25,62 @@ class SecretManager {
+   * @return hash code
    */
-  virtual const SecretSharedPtr staticSecret(const std::string& name) const PURE;
+  static std::string configSourceHash(const envoy::api::v2::core::ConfigSource& config_source) {
```
no impl in `include/`

mangchiandjjoe replied:
Moved to under source/common/secret

jmarantz committed:
tools:  check_format_test to run under CI under 'format' (#3516)

check_format_test is (I think) hard to run under bazel because it must run under the docker image to get the right clang, but bazel is hermetic and I don't know what it would take to call docker from bazel. But it can easily run under CI. In fact it's fast enough to just run it in front of check_format.py itself. Having it in CI and available to run manually via tools/check_format_test.sh should be sufficient to keep check_format.py sane, once more tests are added.

Risk Level: Low
Testing: test/check_format_test.sh --loglevel={0,1,2} and observing CI
Docs Changes: N/A
Release Notes: N/A

Signed-off-by: Joshua Marantz <jmarantz@google.com>

mangchiandjjoe committed:
Addressed review comments

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

mangchiandjjoe committed:
Used std::make_unique

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

mangchiandjjoe committed:
Added SecretManagerUtil class

Signed-off-by: Jae Kim <jaebong.kim@gmail.com>

yuval-k committed:
router: allow path changes in regex routes. (#3537)

Signed-off-by: Yuval Kohavi <yuval.kohavi@gmail.com>

mangchiandjjoe committed:
Merge branch 'dynamic_secret_rebase' into sds_dynamic_secret

mangchiandjjoe committed:
Merge pull request #4 from mangchiandjjoe/sds_dynamic_secret

Sds dynamic secret

mangchiandjjoe committed:
Merge pull request #5 from mangchiandjjoe/dynamic_secret_rebase

Merge pull request #4 from mangchiandjjoe/sds_dynamic_secret

mangchiandjjoe committed:
Revert "Merge pull request #4 from mangchiandjjoe/sds_dynamic_secret"

mangchiandjjoe committed:
Merge pull request #6 from mangchiandjjoe/revert-5-dynamic_secret_rebase

Revert "Merge pull request #4 from mangchiandjjoe/sds_dynamic_secret"

mangchiandjjoe commented:
This branch has been screwed up. I will create another PR for initial review.

mangchiandjjoe commented:
https://github.com/mangchiandjjoe/envoy/pull/8 is the new PR
